Fix bugs in extract_boundary_mesh

starki0815 opened the pull request:
Fixes bugs in extract_boundary_mesh, which caused a broken map between boundary cells and underlying faces of volume mesh as well as inconsistent behavior in 2d and 3d.

In 3d, extract_boundary_mesh swaps vertices to get consistent boundary cell orientations. However, this swapping was so far not taken into account during the assembly of the map between boundary cells and corresponding faces of the volume cells. Fixing the latter required substantial changes (a quick and dirty fix would have been possible in principle, but certainly not clean) to extract_boundary_mesh because the (local) face indices of the volume cells must be known when the map is assembled. Also, the algorithm to do the refinement of the boundary mesh has been modified such that map lookups are no longer necessary. Moreover, vertex swapping has been introduced also for the 2d case in order to make the 2d behavior consistent with the 3d behavior (in previous versions of the function, the normal vector of the boundary mesh pointed always into the domain in 3d, whereas it pointed either into or out of the domain in 2d depending on the situation). Further, an optional parameter has been added to the function, which allows to specify whether the normal vectors should be inward or outward (the default is inward, which has been the behavior of extract_boundary_mesh in 3d in previous versions of the function). Routines have been tested with randomly refined meshes. Fixes #7467.

edit (masterleinad): Clean title and PR summary up.

drwells commented:
Sorry that the review for this is taking so long. I'll try to get to this in the next couple of days.

tjhei commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -1575,6 +1575,9 @@ namespace GridGenerator
    * subset of faces of volume cells for which this function should extract
    * the surface mesh. If left at its default, i.e., if the set is empty, then
    * the function operates on <i>all</i> boundary faces.
+   * @param[in] cell_orientation A boolean determining whether the normal
```
What is the motivation for this parameter? Wouldn't it be acceptable to always have outward facing normals?

tjhei replied:
Second, I would call this parameter ``make_normal_vectors_point_outwards`` or something like that.

starki0815 replied:
The motivation is to retain as much backward compatibility (until now, the normal vector pointed inward in 3d; and either inward or outward in 2d, depending on the situation) as possible while also allowing to choose the often used convention of outward pointing normals. Personally, I really don't care whether the normal vector points inward or outward. So I could on the one hand also imagine to omit the new parameter and let normals always point inwards. On the other hand, I think it's not a big deal to introduce this new parameter and let the user choose.

tjhei replied:
I think tagging additional ``bool`` parameters with default values to the end of a function is a sign for a bad design. I think always using a defined direction is fixing a bug. I would have expected to always get outward facing normals.

drwells replied:
I appreciate the concern for backwards compatibility but I agree with @tjhei: I cannot come up with a situation where we would not want outward-facing normal vectors.

If anyone ever needs this we could simply add `GridTools::invert_normal_orientation` or something like that instead of bundling it with the mesh extraction function.

starki0815 replied:
I interpreted Wolfgang's answer [here](https://groups.google.com/d/msg/dealii/AzHUdaDaifI/q8BJPTCCAgAJ) as that it would be good to stay backward compatible. If not so, I absolutely agree that it's best to eliminate the bool and let the normals always point outward.

drwells replied:
I will mark this as resolved since we have agreed below to have outward normals.

tjhei commented on source/grid/grid_generator.cc:
```diff
@@ -5360,25 +5361,96 @@ namespace GridGenerator
+    std::vector<bool> touched(
+      volume_mesh.get_triangulation().n_vertices(),
+      false); // vector indicating whether a vertex of the volume mesh has
+              // already been visited (necessary to avoid duplicate vertices in
```
We typically put comments in front of expressions. That way they are formatted to use the whole width.

starki0815 replied:
Ok, I often put comments to the side when they refer only to a single line and in front if referring to a whole bunch of comments. But of course, I'll adapt to the dealii style here.

drwells commented on source/grid/grid_generator.cc:
```diff
@@ -5360,25 +5361,96 @@ namespace GridGenerator
     std::map<typename MeshType<dim - 1, spacedim>::cell_iterator,
              typename MeshType<dim, spacedim>::face_iterator>
-      surface_to_volume_mapping;
+      surface_to_volume_mapping; // return of the function
```
Since this is not modified until the end I would just move the declaration down to that point.

starki0815 replied:
Agreed

drwells commented on source/grid/grid_generator.cc:
```diff
@@ -5485,48 +5544,106 @@ namespace GridGenerator
+                    refined_cell->child(swap_matrix[refined_face_number][i3]),
+                    std::make_pair(refined_face->child(i3),
+                                   refined_face_number)));
+              }
```
This would be easier to understand if we used more specific array indices than `i1`, `i2`, and `i3`. How about `cell_n`, `refined_cell_n`, and `child_n`?

drwells replied:
I don't have a strong preference, but good loop variables would make this easier to read.

starki0815 replied:
Agreed, although I don think that it helps much at this point (really descriptive names would have to be prohibitively long here)

drwells commented:
This is going to cause a bunch of tests to fail where we output the mesh. We'll have to tackle that too :)

/run-tests

drwells commented:
One more thing: getting the orientation consistent is a good change to make. Thanks for doing this :)

starki0815 commented on source/grid/grid_generator.cc:
```diff
@@ -5360,25 +5361,96 @@ namespace GridGenerator
+        for (unsigned int i2 = 0; i2 < GeometryInfo<dim - 1>::vertices_per_cell;
+             i2++)
+          swap_matrix[i1][i2] = i2;
+      }
```
Agreed

starki0815 commented on source/grid/grid_generator.cc:
```diff
@@ -5485,48 +5544,106 @@ namespace GridGenerator
+    std::vector<std::pair<
+      typename MeshType<dim - 1, spacedim>::cell_iterator,
+      std::pair<typename MeshType<dim, spacedim>::face_iterator, unsigned int>>>
+      temporary_map_boundaryCell_face; // temporary map between cells on the
```
Agreed

starki0815 commented on source/grid/grid_generator.cc:
```diff
@@ -5360,25 +5361,96 @@ namespace GridGenerator
+    // the entry (i,j) of swap_matrix stores the index of the vertex of
+    // the boundary cell corresponding to the j-th vertex on the i-th face
+    // of the underlying volume cell
+    // if e.g. face 3 of a volume cell is considerd and vertices 1 and 2 of the
```
Changed

starki0815 commented on source/grid/grid_generator.cc:
```diff
@@ -5485,48 +5544,106 @@ namespace GridGenerator
+    // --> check which cells on refinement level i need to
+    // |   be refined -> do refinement (yields cells on level i+1)->--
+    // |                                                             |
+    // --------------------------i+1->i-------------------------------
```
Changed

starki0815 commented:
Concerning the question, whether it could be made simpler: I have also thought about this when writing the code, but don't have a good idea. What can be done for sure is to fix the bug with less changes to the code (actually one only needs a data structure allowing to keep track of face indices). But I think this would neither be very clean nor make the function less complicated to understand. Maybe the part of the code which does the refinement could be written more along the lines of the original implementation. However, this would be associated with a high number of not really necessary map lookups, which I wanted to avoid. Also I think that the original refinement algorithm is not less complicated to understand.

starki0815 commented:
I'll later look into the issue with the tests.

starki0815 commented:
I have now looked into the tests:

Most of them fail because of the modified vertex ordering.

For manifold/polar_manifold_04 there is additionally the bug in the test that only the origin and not the entire z-axis is excluded from the polar manifold.

In several places of the codim_one/extract_boundary_mesh_* tests vertex locations of the faces of the volume mesh and corresponding surface cells are compared. This is not easily possible because vertices are swapped by extract_boundary_mesh. It is only due to some coincidences that this did not cause failed tests in previous versions. In my opinion, the only thing which can be done is a comparison of face centers as the information about which vertices have actually been swapped is not included in the map returned by extract_boundary_mesh [Actually, I think that this loss of information makes the map pretty much useless in practice. For my own purposes I have therefore modified that map such that a face of the volume mesh is not characterized by an iterator to a face, but by the combination of volume cell and face number.].

I have proposed changes to the testsuite to fix the failed tests in the last commit (which in many cases is to update the reference output file). However I have to say that I don't have the time to do an in-depth check of every single test. For most tests, it is fairly obvious that they can't pass with the modified extract_boundary_mesh function and therefore the reference output file must be updated. But this is of course not very satisfactory as it might obscure other bugs of the modified function.

I have also included a new test (codim_one/extract_boundary_mesh_15), which should fail with previous versions of extract_boundary_mesh.

starki0815 commented:
I have now also removed the bool for the normal orientation and modified the function to always produce outward-facing normals. But I would not be surprised if this causes additional failed tests (I have only checked the codim_one/extract_boundary_mesh_* tests so far because I don't have the resources to run the complete testsuite now).

masterleinad commented:
The failing tests are:
```
790 - manifold/tensor_product_manifold_02.debug (Failed)
1547 - codim_one/data_out_03.debug (Failed)
1562 - codim_one/boundary_indicator_01.debug (Failed)
1572 - codim_one/normal_vectors_01.debug (Failed)
1584 - codim_one/mapping_01.debug (Failed)
1585 - codim_one/hanging_nodes_02.debug (Failed)
1587 - codim_one/error_estimator_02.debug (Failed)
1588 - codim_one/hanging_nodes_03.debug (Failed)
1591 - codim_one/hanging_nodes_01.debug (Failed)
1606 - codim_one/transform_01.debug (Failed)
2014 - grid/grid_out_gnuplot_01.debug (Failed)
2048 - grid/grid_out_gnuplot_02.debug (Failed)
```

starki0815 commented:
Yes, these are probably failing because of changing the normal from inward to outward and the associated different order of vertices.

starki0815 commented:
I was just wondering about the following: Is there a reason [other than that set_direction_flag() is not accessible from extract_boundary_mesh()] that extract_boundary_mesh uses vertex swapping to adjust normal directions rather than just using the direction_flag?

The background is the following:

I can think of two situations, where one would like to use extract_boundary_mesh():
(1) Constructing a codim 1 mesh from a volume mesh and then doing computations on the codim 1 mesh;
(2) combining codim 1 finite elements on the boundary and finite elements in the volume if there are variables living in the domain and others living only on the boundary (that is what I do - in fact also on internal interfaces, but this is not much different).

For (1) the present approach of swapping vertices is fine and also the returned map of the function is not really needed.

But for (2) the situation is different because:
(a) the association (w.r.t. order) between children, quadrature points, etc. on the faces of the domain mesh and the corresponding cells of the surface mesh is lost; and to rectify that one really has to know what is going on in extract_boundary_mesh. This makes it difficult to assemble the boundary terms where boundary and volume related  dof's couple.
(b) the map returned by extract_boundary_mesh() does not provide with the information needed. It will be necessary for the user to construct an additional map between faces of the domain mesh and the pair (domain cell, face number).
There is little which can be done about (b) without changing the return of the function.
But if I don't miss anything, adjusting the normals by direction_flag instead by vertex swapping would resolve (a). Also, it would make the function a lot simpler and the whole approach cleaner. The price which has to be paid for that is that either extract_boundary_mesh() is granted access to set_direction_flag() or create_triangulation() is modified in a way that it is possible to specify desired normal orientations for the indivdual cells of a codim 1 mesh ...

What do you think?

starki0815 commented:
What I've also just realized that extract_boundary_mesh won't work in the present form on meshes which have lines with non standard line orientation on the boundary (just try to extract boundary mesh from a moebius mesh with pi/2 twist) ... To get this resolved, one either has to do grid reordering or introduce a line_direction flag also for meshes in 3d, codim 1. Grid reordering would of course again lead to problem (a) from my previous comment.

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -5367,48 +5392,133 @@ namespace GridGenerator
+            index_cells_deepest_level = temporary_map_boundary_cell_face.size();
+            for (unsigned int refined_cell_n = 0;
+                 refined_cell_n < cells_refined.size();
+                 refined_cell_n++)
```
```suggestion
                 ++refined_cell_n)
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -5367,48 +5392,133 @@ namespace GridGenerator
+                      .second.second;
+                for (unsigned int child_n = 0;
+                     child_n < refined_cell->n_children();
+                     child_n++)
```
```suggestion
                     ++child_n)
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -5367,48 +5392,133 @@ namespace GridGenerator
+                  temporary_map_boundary_cell_face
+                    [cells_refined[refined_cell_n]]
+                      .second.first;
+                unsigned int refined_face_number =
```
```suggestion
                const unsigned int refined_face_number =
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -5367,48 +5392,133 @@ namespace GridGenerator
+                 refined_cell_n < cells_refined.size();
+                 refined_cell_n++)
+              {
+                typename MeshType<dim - 1, spacedim>::cell_iterator
```
```suggestion
                const typename MeshType<dim - 1, spacedim>::cell_iterator
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -5367,48 +5392,133 @@ namespace GridGenerator
+                  refined_cell = temporary_map_boundary_cell_face
+                                   [cells_refined[refined_cell_n]]
+                                     .first;
+                typename MeshType<dim, spacedim>::face_iterator refined_face =
```
```suggestion
                const typename MeshType<dim, spacedim>::face_iterator refined_face =
```

starki0815 commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -1579,13 +1579,9 @@ namespace GridGenerator
-   * the cells of the surface mesh.
+   * The order of vertices of surface cells and the corresponding
+   * volume faces may not match in order to ensure that each surface cell is
+   * associated with an outward facing normal.
```
Agreed. But as I have written in my comment above, translating vertices and quadrature points to match quantities on boundary of the volume mesh and on the surface mesh is cumbersome. I am wondering, whether we could do better by setting direction_flag appropriately in extract_boundary_mesh instead of swapping vertices? In particular, I believe that the concept in deal.ii that cells, faces and lines are stored as separate quantities is very reasonable. This way also all codim>0 meshes are in some sense already include as a "subset" in a codim=0 mesh. By swapping vertices, this nice property is  destroyed more or less ...

bangerth replied:
I suspect you can always come up with cases where you need some kind of translation between vertices. The problem is that we want to impose certain properties in the meshes we have and asking for too many properties to be true at the same time may make the set of meshes we can represent zero. I'm not saying that what you suggest is not possible, but I would not be surprised if you can't always do it that way.

starki0815 replied:
I am not sure whether I really understand what you say. The only case I can think of, where the approach I am suggesting won't work is if the there are lines with non standard orientation on the boundary of a volume mesh. But this doesn't work with the current implementation of extract_boundary_mesh either (with the current implementation, it could probably made work by doing reordering of the grid if lines with non standard orientation are present). With the approach I am suggesting, one would have to allow line_direction to be false for a 3d codim 1 mesh - I suspect that it would indeed be difficult to make this work. Is this what you mean? Or is it maybe not possible to arbitrarily set the direction_flag for a 3d codim 1 mesh (i.e., are there constraints on how the direction_flag of neighboring cells is set)?

bangerth replied:
You catch me flatfooted. There are probably only 3 or so people who really understood how these direction flags work. I am one of those, but I don't remember the details of it all any more. So if you think it can work, then I would say try it out!

I am an incrementalist. So one of the possible ways forward would be to finish the patch in its current form (assuming there is anything to finish -- I haven't looked), get it merged, and then do as you suggest with the direction flags. This way, there is a known and working state in the repository that's already an improvement over what's in 9.0. Then we can see if your idea pans out: if it does, great, which we can check with a few additional tests; if it doesn't, then at least we can stick with the state of the current PR which by then is in the repository.

starki0815 replied:
Good, I think the approach you propose is reasonable. Let's do it like this. For now it remains only to fix some tests which are broken by the patch.

bangerth replied:
OK. How do we go about this then? If you are confident that the patch is correct -- which I think is credible -- can you just update the test outputs?

starki0815 replied:
Most of the tests should fail either because normal vectors are output or the mesh is written to a file. It's fairly easy to check that this is indeed the case and this is what I'll definitely do plus a few plausibility checks and visual inspection of the meshes produced. I have already fixed 20 or so broken tests this way and I'll do it the same with the remaining 12. Doing real in depth checking of every single test would take ages ...

bangerth replied:
I think it's ok if a test passes the smell check. If the changes in 20 tests are all to be expected, the likelihood that the remaining 12 just follow the same pattern is quite high and I don't see why you need to look at every single one in great detail -- just update the output files.

starki0815 commented on source/grid/grid_generator.cc:
```diff
@@ -5367,48 +5392,133 @@ namespace GridGenerator
+    std::vector<std::pair<
+      typename MeshType<dim - 1, spacedim>::cell_iterator,
+      std::pair<typename MeshType<dim, spacedim>::face_iterator, unsigned int>>>
+      temporary_map_boundary_cell_face;
```
Hm, I have written the refinement algorithm in a way that I don't have to search in any container like temporary_map_boundary_cell_face if this is what you mean. So, I can't see an advantage of replacing the vector by a map. But I think there are many ways how the refinement process can be performed ...

bangerth replied:
If you don't need to search, then a `std::vector<std::pair<...>>` is the right data structure. I was just curious how you use the map, given that it's called "map" :-)

masterleinad commented:
Can you please squash your commits to a few self-contained ones?

starki0815 commented:
I have squashed the commits and hope everything is working now.

bangerth commented:
Thanks, @starki0815, for the patch and all of the work associated! I hope it's the first in a long list of patches from you! My apologies also for dragging this out over 6 weeks :-(

Feel free to go about with your suggestion for an improvement of the function. Let's hope we can all remember how these things work with all of the associated direction and other flags :-)
